feat: Adds BS5 modal components

mavoIn opened the pull request:
As discussed in #233  I added a modal component. It works like the cards component. Basic documentation added as well.

fsbraun commented:
So coool! 😎 I will look into it! Can you get the pre-commit stuff sorted out meanwhile?

mavoIn commented:
Hi @fsbraun, it is my first pull request ever. Maybe you have to look over the code twice. 😇
I tested it locally. Meanwhile I use the forked version. 
Best mavoIn

fsbraun commented:
Hey @mavoIn ! Can I ask you to do the following things:
* `pip install ruff` (if you haven't done so already) and `ruff check --fix djangocms_frontend`
* Add tests in the `tests` folder. Please feel free to take similar tests as a blueprint. The tests are not very sophisticated but just check if the model and the rendering is working.

I hopefully will be able to test the modal in a pet project tomorrow.

mavoIn commented:
HI @fsbraun , 
Unfortunately, I have only been able to prepare a few test files for you. They are not complete.
I am currently busy learning more. :D

fsbraun reviewed (changes requested):
I think I need to better understand how to use the modal? Maybe you can give an example in the docs?

fsbraun commented:
Also, can you take a look at the import issue showing up in tests? And the migrations seem not up-to-date. You may update the initial migration - they are DB no-ops anyways.

mavoIn commented on docs/source/components.rst:
```diff
@@ -280,6 +280,21 @@ Media component
+Modal component
+*******************
+
+The modal component shows content inside an modal.
```
added

mavoIn commented:
Hello @fsbraun, thanks a lot for all your advices.
I added the missing migrations and updated the documentation. 

There are problems importing the tests because I was only able to insert a code framework from the collapse example. Please let me know so that you can fix the tests or I might have to ask friends for further support.

Best mavoIn

fsbraun commented on tests/modal/test_models.py:
```diff
@@ -0,0 +1,32 @@
+
+
+class ModalModelTestCase(TestCase):
+   def test_Modal_instance(self):
```
```suggestion
    def test_Modal_instance(self):
```
There's an indentation error here.

fsbraun commented:
@mavoIn I see you have updated quite a few things! Thank you. Still, the tests are failing. Try running them locally:  `python ./run_tests.py`. There seems to be an import issue, an attribute issue and maybe some tweaking for the migrations required.

fsbraun commented on djangocms_frontend/contrib/modal/forms.py:
```diff
@@ -0,0 +1,167 @@
+    )
+
+    modal_static = forms.BooleanField(
+        label=_("Static"),
```
```suggestion
        label=_("Static backdrop"),
```
I assume this refers to https://getbootstrap.com/docs/5.2/components/modal/#static-backdrop

fsbraun commented on djangocms_frontend/contrib/modal/forms.py:
```diff
@@ -0,0 +1,167 @@
+    modal_static = forms.BooleanField(
+        label=_("Static"),
+        required=False,
+        help_text=_("Disable scrolling in the container."),
```
```suggestion
        help_text=_("If selected, the modal will not close when clicking outside of it."),
```

fsbraun commented on djangocms_frontend/contrib/modal/cms_plugins.py:
```diff
@@ -0,0 +1,119 @@
+    form = forms.ModalTriggerForm
+    allow_children = True
+    parent_classes = [
+        "CardPlugin"
```
I am not sure why a modal trigger should be part of a card? I am having difficulties imagining a static button or other trigger that should trigger a modal. Can you give a use-case example? Things that come to my mind are typically related to a POST request, e.g., form submissions (Say, for confirmations)

mavoIn replied:
You're right, @fsbraun. The parent class should be "ModalPlugin", not CardPlugin. 
The trigger for the collapse could be everywhere on the cms page. I thought that the ModalTrigger should be inside the ModalPlugin to force an organised page tree. Or do you think it is better to remove this dependency?

fsbraun commented on djangocms_frontend/contrib/modal/templates/djangocms_frontend/bootstrap5/modal-inner.html:
```diff
@@ -0,0 +1,11 @@
+        {% with forloop as parentloop %}{% render_plugin plugin %}{% endwith %}
+    {% endfor %}
+{% endwith %}
+{% if instance.inner_type == "modal-header" %}
```
Should a close button not be optional?

mavoIn replied:
Hi @fsbraun, you are right. We should add an option to show  an x in the header, a close button  in the footer, both or none. Is this what you expect?

mavoIn commented:
> I have some concerns about the structure:
> 
> * Modal triggers should be possible anywhere on a page
> * The Modal plugin is not needed as far as I can see. I think the `<div class="modal"></div>` can go into the modal container.
> * I still would need to understand a use case with a static trigger. Can you give an exmaple?

Hi @fsbraun, what do you exactly mean with static trigger? 
Do you mean the requirement of the  trigger inside the modal component? 
From my perspective i want to prevent user errors by using the loosely coupled modal trigger like it is done in the collapse example.  
Best mavo
